### ISSUE TITLE
Import alias not working correctly in Strapi typescript project

### ISSUE THREAD

**Issue by Paiman-Rasoli** [strapi/strapi#18952]
### Required System information

- Node.js version: 16.20.2
- NPM version: 8.19.4
- Strapi version: 4.13.1
- Database: _
- Operating system: windows 11
- Is your project Javascript or Typescript: Typescript

### Describe the bug

I defined some import alias path in my `tsconfig.json` file and typescript intellisense know the files and I can navigate to the files that were imported by an alias, however when I run `npm run develop` I get an error that the module `alias-name` is not found.

`tsconfig.json`
```json
{
  "extends": "@strapi/typescript-utils/tsconfigs/server",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": ".",
    "baseUrl": ".",
    "paths": {
      "@custom/plugins": ["custom/plugins/index.ts"],
    }
  },
  "include": ["./", "./**/*.ts", "./**/*.js", "src/**/*.json"],
  "exclude": [
    "node_modules/",
    "build/",
    "dist/",
    ".cache/",
    ".tmp/",
    "src/admin/",
    "**/*.test.*",
    "src/plugins/**"
  ]
}

```
the error: Cannot find module '@custom/plugins'

and this is the build file that I imported file by the alias
```javascript
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugins_1 = require("@custom/plugins");
const constants_1 = require("../../../../../utils/constants");
const profile_1 = require("../../../../../utils/services/profile");
```
I guess the error is when compiling to `js`  it does not replace the alias import to its relative path.
Thanks in advance!

**Comment by Ghustavh97** [comment]
Having the same issue as well, using relative paths for now.

**Comment by EsraaSaad86** [comment]
Did you try to add your module alias to your package.json file? As after building dist files is JS so it won't be able to work with tsconfig files.
I think you will just need to add something like this to you package.json
```
"_moduleAliases": {
"@custom": "./dist/src/custom" //Just make sure to add the path to your plugin inside dist folder
}
```
then you can use this path like this
import {} from "@custom/plugins/index" this should refer to your index file inside the plugins
As I don't think you need to add /index.ts to your path so you can easily configure it when it comes to dist folder as with dist folder it will be index.js

**Comment by Frost-on-Web** [comment]
Hi,
Having the same issue here trying to have a folder at the root : "./lib" (with some "interfaces" classes into).

So I defined into the `tsconfig.json` :
```
"paths": {
   "@lib/*": ["./lib/*"]
}
```

and using it into a custom controller :
```
import { zPurchaseOrder } from "@lib/shop-interfaces/PaymentFormTypes";
```

But having always the same error when I "yarn dev" :
```
Cannot find module '@lib/shop-interfaces/PaymentFormTypes'  
```

And trying the "_moduleAliases" upper solution is not a solution for me.

%%%%%%%%%%%%%%%%%

For information, my "./dist" output folder seems to have the good structure.
But, I found the controller such :
```
const PaymentFormTypes_1 = require("@lib/shop-interfaces/PaymentFormTypes");
```

I'm not an expert but I would thought than into the "./dist" folder, the tsconfig aliases might be converted, aren't ?

%%%%%%%%%%%%%%%%%

It's not a big messy issue but it will be appreciate to simplify development.
Have a great day \o/

**Comment by ridgekuhn** [comment]
Hello, I'm new to Strapi so please forgive any ignorance about it, but I've just run into this issue and can explain a little bit, so hopefully this helps someone.

The `paths` option in `tsconfig` only informs Typescript that our project uses path aliases created by some other means; eg, by [module-alias](https://www.npmjs.com/package/module-alias) (used for defining path aliases in package.json as mentioned in a comment above), or confusingly, [tsconfig-paths](https://www.npmjs.com/package/tsconfig-paths). Simply setting paths in `tsconfig` [does not cause Typescript to replace paths](https://www.typescriptlang.org/tsconfig#paths) in code emitted by the compiler, so Strapi is actually behaving as expected here. However, Strapi makes the normal `tsconfig` paths workflow impossible, so an upstream solution to the problem would be nice.

Normally, you might use `tsconfig-paths` during development, passing the module to `node-dev`/`ts-node-dev` as a requirement, eg, `$ ts-node-dev -r tsconfig-paths/register`. `tsconfig-paths` then replaces Node’s runtime module resolver with one that can resolve the `tsconfig` paths. In this scenario, you are either running the Typescript code directly with `ts-node`, or running from a compiled directory with `node`. Either way, the directory structure relative to the parent process' working directory matches what's in your `tsconfig` paths. For production, you might compile with `tsc`, followed by something like [tsconfig-replace-paths](https://www.npmjs.com/package/tsconfig-replace-paths) to actually replace all the path aliases with absolute ones in the compiled code, saving the production app the extra runtime dependency and having to waste time/resources doing path resolution.

The problem with Strapi is that for both development and production runtime, the project is first compiled to `dist/`, and the parent process runs the compiled code from there, but with the project root as the process' working directory. So, even if you were able to use your `tsconfig` paths, they would be wrong because they would lack the `dist/` prefix. Of course, we can’t just add the `dist/` prefix to our `tsconfig` paths, because then they would no longer work with Typescript.

It seems Strapi implements a [custom compiler](https://github.com/strapi/strapi/blob/bbb5d9ccd74376825743893bf92696f0e4e6cfb7/packages/utils/typescript/lib/compilers/basic.js#L29) rather than using something like `tsc`. When `ts.createProgram()` receives the `options` argument, all of our `tsconfig` settings like `baseUrl` and `paths` are available, so I think the cleanest solution would be if Strapi’s compiler could take the path settings, apply the `dist/` prefix, and then do the path replacement after compilation, like a bundler might do. Speaking of bundlers, plugins like [vite-tsconfig-paths](https://www.npmjs.com/package/vite-tsconfig-paths) and [tsconfig-paths-webpack-plugin](https://www.npmjs.com/package/tsconfig-paths-webpack-plugin) are available. I see that both Webpack and Vite are available to Strapi, but it appears that either is only used for building the admin frontend?

tl;dr, the workaround is to use `module-alias`, it allows us to define path aliases independent of Typescript, so we can define the paths with the `dist/` prefix in package.json (or inline when we import the module). Typescript is happy during development because our paths in `tsconfig` point to Typescript files, and Node is happy at runtime because `module-alias` points Node's resolver to the `dist/` dir. Unfortunately, even though we can pass a custom `tsconfig` to `tsconfig-replace-paths`, it doesn't seem like there's any combination of settings that will allow it to prefix replaced paths with `dist/`, probably because Typescript itself assumes you would never run your program from a directory above the compiled code it emits, as Strapi does. So, our production app is stuck with the extra path resolver module dependency and having to waste resources doing path resolution, and our codebase is stuck with a mess of config files that we'll have to keep manually synced, but at least the paths are portable.

**Comment by seleckis** [comment]
Thank you, @ridgekuhn for explanation. Let's see what Developers Experience Squad can do with it.

**Comment by mjcurciojr** [comment]
Issue is still present, configuring path aliases in .tsconfig still does not work correctly even in strapi 5. Asking users to install module-alias is really not a suitable workaround, IMO
